text/0000-edition-2021-and-beyond.md

```diff
@@ -0,0 +1,194 @@
+
+This RFC proposes a specific answer to those questions:
+
+* We will do new Rust editions on a regular, three-year cadence.
+    * We will roll out an edition regardless of whether there are breaking changes.
+* Unlike Rust 2018, we will avoid using editions as a "deadline" to tie together high-priority projects.
+    * Instead, we embrace the train model, but editions are effectively a "somewhat bigger release", giving us an opportunity to give an overview of all the work that has landed over the previous three years.
+* We specify a cadence for Edition lints.
+    * "Edition idiom" lints for Edition N will warn for editions before N, and become "deny by default" in Edition N.
+    * Since it would be disruptive to introduce deny-by-default lints for Rust 2018 now, the Rust 2018 lints are repurposed into Rust 2021 Edition lints.
+* We specify a policy on reserving keywords and other prospective changes.
+    * In short, reserving keywords is allowed only as part of an active project group.
+
+## Expected nature of editions to come
+
+We believe the Rust 2018 was somewhat exceptional in that it introduced changes to the module system that affected virtually every crate, even if those changes were almost completely automated. We expect that the changes introduced by most editions will be much more modest and discrete, more analogous to `async fn` (which simply introduced the `async` keyword), or the changes proposed by [RFC 2229] (which tweaks the way that closure captures work to make them more precise).
+
+The "size" of changes to expect is important, because they help inform the best way to ship editions. Since we expect most changes to be relatively small, we would like to design a system that allows us to judge those changes individually, without having to justify an edition by having a large number of changes combined together. Moreover, we'd like to have editions happening on a predictable cadence, so that we can take that cadence into account when designing and implementing features (i.e., so that we can try to tackle changes that may require migrations earlier, to give plenty of time).
+
+## Key ideas of edition do not change
+
+Just as with Rust 2018, we are firmly committed to the core concepts of an edition:
+
+* Crates using older editions continues to compile in the newer
+  compiler, potentially with warnings.
+* Crates using different editions can interoperate, and people can
+  upgrade to newer editions on their own schedule.
+* Code that compiles without a warning on Edition N should also
+  compile on Edition N + 1.
+* Migration between editions should generally be automated.
+* Editions make "skin-deep" changes, with all editions ultimately
+  compiling to a single common representation.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We use this section to try and convey the story that average users will need to understand.
+
+## What is a Rust edition?
```

Comment:
To me the paragraphs here don't really answer the question _"What __is__ a Rust edition?"_.
It is briefly touched in the last pragraph but not in detail. So it is just that year's version that allows us to introduce features that would be impossible (so are  these breaking changes I assume).

Then the question remains, if there are no breaking changes, why do we want a new edition? Wouldn't that cause confusion? For example, do we need to still set the edition number if there are 0 breaking changes?

Reply:
So that you can always know to set your edition value to 2015+3x and you'll get a valid number.

Reply:
We could should probably start here with: _"A rust edition is ..."_ and maybe also mention what it is not, and the other paragraphs regarding name, cadence and opportunity to look back and celebrate can follow.

Reply:
I guess my main worry (besides the lack of a definition) is the dual meaning of an edition, it can be a breaking change or can be a non-breaking changes at the same time. 

This is the main reason that makes me ask all the questions above and I am sure that other users of rust may have the same dificulties. 

I am more of a fan where things have a clean meaning in all situations:
- a compiler update is always a non-braking change (at least not intentional)
- rust editions are breaking changes that give you some help in porting and also make it worth to invest the effort due to new features + compatibility + auto-changes from tools + etc.

Reply:
> I was looking at the patch and noticed that the change is just changing from 2018 to 2021. I wonder what is the point of editions if there are no breaking change? Why would one change from 2018 to 2021 if it is the same thing? If it is the same thing then most likely some crates will be 2018 and some will be 2021, will that slow down the compile time? But from the RFC looks like we have breaking change.

I posted this in the PR which I should ask here. I am thinking that if there are no breaking change, means that people may not change to the newer edition since it is the same. Questions

- Why would one change the edition to a newer one which is the same?
- Is it necessary to have a new edition where the new edition only breaking change is changing 2018 to 2021?
- Will it slow down the compilation time?